fix(ivy): return new ViewRef when detaching view

crisbeto opened the pull request:
When detaching a view by its index via `ViewContainerRef.detach(index)`, in `ViewEngine` we used to return a new `ViewRef` ([for reference](https://github.com/angular/angular/blob/master/packages/core/src/view/refs.ts#L227)), however in Ivy we return the same `ViewRef` which means that its internal `_viewContainerRef` is never reset and we'll throw an error if the consumer tried to attach it to the `ApplicationRef`. These changes return a new `ViewRef` in order to match the original behavior.

These changes also add the same errors as `ViewEngine` when attempting to attach a view that is attached already. This was the original goal of this PR, however it ended up uncovering the issues with the `ViewRef`.

pkozlowski-opensource commented on packages/core/src/render3/view_ref.ts:
```diff
@@ -258,11 +258,21 @@ export class ViewRef<T> implements viewEngine_EmbeddedViewRef<T>, viewEngine_Int
 
-  attachToViewContainerRef(vcRef: viewEngine_ViewContainerRef) { this._viewContainerRef = vcRef; }
+  attachToViewContainerRef(vcRef: viewEngine_ViewContainerRef) {
+    if (this._appRef) {
```
I think that we should guard this check and error with the dev mode so this code can be tree-shaken away in the production mode.

pkozlowski-opensource commented on packages/core/src/render3/view_ref.ts:
```diff
@@ -258,11 +258,21 @@ export class ViewRef<T> implements viewEngine_EmbeddedViewRef<T>, viewEngine_Int
 
-  attachToAppRef(appRef: ApplicationRef) { this._appRef = appRef; }
+  attachToAppRef(appRef: ApplicationRef) {
+    if (this._viewContainerRef) {
```
I think that we should guard this check and error with the dev mode so this code can be tree-shaken away in the production mode.

pkozlowski-opensource reviewed (changes requested):
Could you please guard those error messages with the dev mode so we are tree shaking it in the production mode?

crisbeto committed:
fix(ivy): return new ViewRef when detaching view

When detaching a view by its index via `ViewContainerRef.detach(index)`, in `ViewEngine` we used to return a new `ViewRef` ([for reference](https://github.com/angular/angular/blob/master/packages/core/src/view/refs.ts#L227)), however in Ivy we return the same `ViewRef` which means that its internal `_viewContainerRef` is never reset and we'll throw an error if the consumer tried to attach it to the `ApplicationRef`. These changes return a new `ViewRef` in order to match the original behavior.

These changes also add the same errors as `ViewEngine` when attempting to attach a view that is attached already. This was the original goal of this PR, however it ended up uncovering the issues with the `ViewRef`.

crisbeto commented:
@pkozlowski-opensource I addressed the comment on the docs. Regarding guarding these messages to dev mode, if we did it, we'd be deviating from the ViewEngine behavior. See https://github.com/angular/angular/blob/master/packages/core/src/view/refs.ts#L294.

mary-poppins commented:
You can preview 0ce51c0 at https://pr27437-0ce51c0.ngbuilds.io/.

kara reviewed (approved):
LGTM

pkozlowski-opensource reviewed (approved):
OK, let's keep error messages consistent with the view engine for now.

kara commented:
[presubmit](http://test/OCL:224008729:BASE:224008775:1543950077168:3bfb882)
